### PR TITLE
new asdf extension API

### LINKED ISSUE #368: rename schema tags to asdf:// format
breakout from #123

rework tag naming structure to match https://asdf.readthedocs.io/en/al-503-document-2.8-features/asdf/extending/uris.html#tags

### PR DESCRIPTION
**Pull request by CagtayFabry**
# Notes on working with the new extension

## asdf documentation
read https://asdf.readthedocs.io/en/latest/index.html#extending-asdf

## file and directory layout for new weldx asdf extension:
```
weldx
|- asdf
  |- constants.py # schema files path and URI strings
  |- types.py # definition of base WeldxConverter class to inherit
  |- extension.py # setup of schema resource mapping and weldx extension
  |- _extension.py # legacy extension code used to register the weldx validators
|- manifest
  |- weldx-1.0.0.yaml # the new manifest file
|- schemas
  |- # all schema definition files, formerly in weldx/asdf/schemas
|- tags
  |- # all WeldxConvert implementations, formerly in weldx/asdf/tags
```

## Manifest
I have added a script in `weldx.asdf.cli` to read the existing manifest and update it in place with all currently supported tags
A careful look at the git diff is certainly required

## regex patterns
useful regex patterns for conversion in PyCharm are listed in #479

## missing features in new API
the following features are not yet supported in the new API
- control of custom array storage in `to_yaml_tree` methods, this feature in LCS has been disabled for now
- custom validators can not be registered via the new extension API so far, they have been implemented using the legacy API https://github.com/asdf-format/asdf/issues/1012

### support for legacy weldx files/tags
it is  possible to open legacy weldx files with the following:
- [x] add legacy tag name to list (via metaclass)
- [x] add legacy tag-schema relations to manifest (manually via regex replace)
- [x] remove `tag` lines in all schema definitions (not needed any longer)
- [x] register the legacy extension name (might not be needed but avoids warnings?)
- [x] test with existing weldx files
- [x] `*_yaml_tree` functions are no longer classmethods

### open issues and API changes to consider/redo
- [x] check support for version `1.*` with groove types
- [x] correctly register entry points (for conda build! and asdf pytest!!)
- [x] enable `show_asdf_header()` in `WeldxFile.test_custom_schema` again
- [x] rework manifest collection helper script
- [x] unify custom shape validation integration
- [x] remove legacy code

## Related Issues

Closes #368, #370, #371, #382, #86, #482

## Checks

- [x] updated CHANGELOG.md
- [x] updated tests
- [x] updated doc/
- [x] update example/tutorial notebooks

### CODE REVIEW

**Comment by CagtayFabry** [comment]
I have marked this 'ready for review' only to enable the build CI steps

**Comment by CagtayFabry** [comment]
@vhirtham this should be ready now finally

unfortunately putting the schema files directly into `weldx/schemas` seems to cause an error with the asdf pytest plugin so I have kept the old subfolders
It is a lot easier to manage and change the folders using the paths defined in `wedx.asdf.constants` now so we can change that at some point again

I will update the CHANGELOG later and add some useful tips in the discussion section

**Comment by CagtayFabry** [comment]
we will also need to update the [Quality Standard template](https://github.com/BAMWelDX/quality-standard-template) to the nee URI names and updated schemas, maybe you can have a look @vhirtham ?

**Review by vhirtham — APPROVED** [review]
Didn't recheck every schema for obvious reasons 😉 The tests are passing, so it should be fine. Have a look at my two small comments and feel free to merge afterward.

File: weldx/asdf/util.py [hunk 1]
```diff
@@ -327,43 +338,88 @@ def dataclass_serialization_class(
...
+
+
+def get_converter_for_uri(uri: str) -> Union[type, bool]:
+    """Get the converter class that handles a given tag."""
+    converters = [s for s in WeldxConverter.__subclasses__() if uri_match(s.tags, uri)]
+    if len(converters) > 1:
+        warn(f"Found more than one converter class for uri {uri}", UserWarning)
+    if converters:
+        return converters[0]
+    return False
```
[the remark this answers is not in the document]
Comment (CagtayFabry) [reply weldx/asdf/util.py:409]: `None` probably makes more sense here, true :)

**Comment by CagtayFabry** [comment]
I will look through it again to update the changelog and merge after, code looks fine :)

**Pull request merged**